openupgradelib/openupgrade.py

```diff
@@ -1876,6 +1878,50 @@ def wrapped_function(cr, version):
     return wrap
 
 
+def progress(index=1, prefix=None):
+    """This is a decorator for any sub function called in an OpenUpgrade script
+    (pre, post or end migration script).
+
+    Decorate functions that may take time. If a function is decorated, we
+    provide an iterable argument that will be looped on and call the original
+    function, while a progress bar will be displayed.
+
+    Function: 28% (152 of 529) I#####              I Elapsed Time: 0:00:03 ETA: 0:01:32
+
+    :param index: Index of the argument to be used as iterable. Default to the
+      second argument. It will pass each of the elements of the iterable in the
+      same place.
+    :param title: Optional title for prefixing the progress bar. If not
+      specified, the function name will be used.
+
+    Typical use::
+
+        @openupgrade.progress()
+        def migrate_some_stuff(env, record)
```

Comment:
That can't be possible, Sylvain, as this needs to control each step of the loop for the progress bar. Think on this like `@api.one` decorator, in which `self` is only one record. You pass the whole recordset, but what gets to your method is only one. Here I took care of putting the proper nomenclature of the arguments for not confusing.

The only way to get what you want is to intercept things into another level, like initially @ypapouin did in his PR #206, but that means that we need an specific decorator for each kind of operation (his one was for a fetchall), and lose the general purpose. We can override the iterator and create a wrapper, but this still is limited, as you can't have several nested loops, so it's not a lot of problem to change this. Now it seems a bit dirty as applied over a method that previously was not "progressified", but when writing new methods, there won't be diff and even the cyclomatic complexity will be lower, reducing one indentation level.

Reply:
Hi @pedrobaeza. 

thanks for your answer.
In a first sight, I was thinking to replace ``args_copy[index] = elem`` by ``args_copy[index] = [elem]``. It can do the job.

- It's not perfect and should ask some minor changes for dict but it avoid huge diff. (``dict.items() can not be used inside and should be replaced``)
- Also, if using progress bar generates side bad effects for any reasons (performance, in the log, or whatever), it could be great to disable / remove it easily)
- if using the progress bar requires to change the code, it will slow down the use of this new decorator for the current migration scripts.
- when writing function, it is counter intuitive to call a function with a list, and in the function, to have a single element to consume.

In the other hand, 

- I like the reduction of the complexity you're talking about 
- indeed it will not be a problem for new function

I don't know ! I have mixed feelings on that point. Maybe other openupgrader will will have additional opinions.

Any way, not a blocking point. I think that introducing progressbar is a step in the right direction.

Reply:
I sincerely don't think passing a list of one element is the way to go. It's a bit antinatural and only for these cases where we adapt existing methods to reduce the diff, but it's less readable and put in doubt what the decorator does IMO.

Reply:
> when writing function, it is counter intuitive to call a function with a list, and in the function, to have a single element to consume.

This is done in other places like `migrate(cr, version)` that gets converted to `migrate(env, version)`.